rust/ethers-prometheus/src/lib.rs

```diff
@@ -0,0 +1,500 @@
+                    // Okay, so Ether is not a token, but whatever, close enough.
+                    let balance = u256_as_scaled_f64(balance, 18);
+                    trace!("Wallet {wallet_name} ({wallet_addr_str}) on chain {chain} balance is {balance}ETH");
+                    wallet_balance_metric
+                        .with(&hashmap! {
+                        "chain" => chain,
+                        "wallet_address" => wallet_addr_str.as_str(),
+                        "wallet_name" => wallet_name,
+                        "token_address" => "none",
+                        "token_symbol" => "ETH",
+                        "token_name" => "Ether"
+                    }).set(balance)
+                },
+                Err(e) => warn!("Metric update failed for wallet {wallet_name} ({wallet_addr_str}) on chain {chain} balance for Ether; {e}")
+            }
+            for (token_addr, token) in data.tokens.iter() {
+                let token_addr_str = token_addr.to_string();
+                let balance = match Erc20::new(*token_addr, client.clone())
+                    .balance_of(*wallet_addr)
+                    .call()
+                    .await
+                {
+                    Ok(b) => u256_as_scaled_f64(b, token.decimals),
+                    Err(e) => {
+                        warn!("Metric update failed for wallet {wallet_name} ({wallet_addr_str}) on chain {chain} balance for {name}; {e}", name=token.name);
+                        continue;
+                    }
+                };
+                trace!("Wallet {wallet_name} ({wallet_addr_str}) on chain {chain} balance is {balance}{}", token.symbol);
+                wallet_balance_metric
+                    .with(&hashmap! {
+                        "chain" => chain,
+                        "wallet_address" => wallet_addr_str.as_str(),
+                        "wallet_name" => wallet_name,
+                        "token_address" => token_addr_str.as_str(),
+                        "token_symbol" => token.symbol.as_str(),
+                        "token_name" => token.symbol.as_str()
+                    })
+                    .set(balance);
+            }
+        }
+    }
+}
+
+impl<M: Middleware> Debug for PrometheusMiddleware<M> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "PrometheusMiddleware({:?})", self.inner)
+    }
+}
+
+/// Get the metrics appropriate chain name from the chain ID.
+pub fn metrics_chain_name(chain_id: Option<u64>) -> String {
+    if let Some(chain_id) = chain_id {
+        if let Ok(chain) = Chain::try_from(chain_id) {
+            format!("{chain}")
+        } else {
+            format!("{chain_id}")
+        }
+    } else {
+        "unknown".into()
+    }
+}
+
+/// Convert a u256 scaled integer value into the corresponding f64 value.
+fn u256_as_scaled_f64(value: U256, decimals: u8) -> f64 {
+    value.to_f64_lossy() / (10u64.pow(decimals as u32) as f64)
```

Comment:
whered `to_f64_lossy` move to? for some reason can't find it

Reply:
Had to enable a feature flag for a sub dependency to access it.
https://github.com/paritytech/parity-common/blob/master/primitive-types/src/fp_conversion.rs#L35